### PR TITLE
Dynamically add java handlers to velocity template

### PR DESCRIPTION
**Pull request by lahiru823**

### CODE REVIEW

File: components/appmgt/org.wso2.carbon.appmgt.api/src/main/java/org/wso2/carbon/appmgt/api/model/JavaPolicy.java [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+package org.wso2.carbon.appmgt.api.model;
+
+import org.json.simple.JSONArray;
+
+public class JavaPolicy {
+    private Integer policyID;
+    private String policyName;
+    private String fullQualifiName; //full qualified name
+    private Integer order; //display order
+    private JSONArray properties;
```
Comment (rushmin) [inline components/appmgt/org.wso2.carbon.appmgt.api/src/main/java/org/wso2/carbon/appmgt/api/model/JavaPolicy.java:28]: Replace JSONArray with a vanilla List.

File: components/appmgt/org.wso2.carbon.appmgt.impl/src/main/java/org/wso2/carbon/appmgt/impl/APIProviderImpl.java [hunk 2]
```diff
@@ -875,45 +877,62 @@ private boolean isAPIPublished(WebApp api) throws AppManagementException {
...
+     * @param api :WebApp class which contains details about web applications
+     * @return :handlers list with properties to be applied
+     * @throws AppManagementException on error
+     */
+    private APITemplateBuilder getAPITemplateBuilder(WebApp api) throws AppManagementException {
+        APITemplateBuilderImpl vtb = new APITemplateBuilderImpl(api);
+
+        //List of JavaPolicy class which contains policy related details
+        List<JavaPolicy> policies = new ArrayList<JavaPolicy>();
+        //contains properties related to all the policies
```
Comment (rushmin) [inline components/appmgt/org.wso2.carbon.appmgt.impl/src/main/java/org/wso2/carbon/appmgt/impl/APIProviderImpl.java:892]: Comments should start with a capital letter.

File: components/appmgt/org.wso2.carbon.appmgt.impl/src/main/java/org/wso2/carbon/appmgt/impl/APIProviderImpl.java [hunk 3]
```diff
@@ -875,45 +877,62 @@ private boolean isAPIPublished(WebApp api) throws AppManagementException {
...
+        int counterPolicies; //counter :policies
+        int counterProperties; //counter :properties
+        String propKey; //Property Key name
+        String propVal; //Property value
+
+        try {
+            //fetch all the java policy handlers details which need to be included to synapse api config file
+            policies = appMDAO.getMappedJavaPolicyList(api.getUUID(),true);
+            //loop through each policy
+            for (counterPolicies = 0; counterPolicies < policies.size(); counterPolicies++) {
```
Comment (rushmin) [inline components/appmgt/org.wso2.carbon.appmgt.impl/src/main/java/org/wso2/carbon/appmgt/impl/APIProviderImpl.java:905]: 'counterPolicies' is not a meaningful name.

File: components/appmgt/org.wso2.carbon.appmgt.impl/src/main/java/org/wso2/carbon/appmgt/impl/dao/AppMDAO.java [hunk 4]
```diff
@@ -4505,6 +4511,45 @@ private void updatePolicyGroups(int applicationId, Object[] policyPartialList,
...
+			saveJavaPolicyMappings(connection, applicationId, javaPolicyList);
+
+		} catch (SQLException e) {
+			/*
+			In the code im using a single SQL connection passed from the parent function so I'm logging the error here
+			and throwing the SQLException so the connection will be disposed by the parent function.
+			*/
+			log.error("Error while deleting XACML policy partial mappings for webapp : " +
+					applicationId, e);
+			throw e;
```
Comment (rushmin) [inline components/appmgt/org.wso2.carbon.appmgt.impl/src/main/java/org/wso2/carbon/appmgt/impl/dao/AppMDAO.java:4547]: Do not re-throw the SQLException.

File: components/appmgt/org.wso2.carbon.appmgt.impl/src/main/java/org/wso2/carbon/appmgt/impl/APIProviderImpl.java [hunk 5]
```diff
@@ -875,45 +877,62 @@ private boolean isAPIPublished(WebApp api) throws AppManagementException {
...
-	}
+    /**
+     * This method dynamically returns the mandatory and selected java policy handlers list for given app
+     *
+     * @param api :WebApp class which contains details about web applications
+     * @return :handlers list with properties to be applied
+     * @throws AppManagementException on error
+     */
+    private APITemplateBuilder getAPITemplateBuilder(WebApp api) throws AppManagementException {
+        APITemplateBuilderImpl vtb = new APITemplateBuilderImpl(api);
```
Comment (dinusha-dilrukshi) [inline components/appmgt/org.wso2.carbon.appmgt.impl/src/main/java/org/wso2/carbon/appmgt/impl/APIProviderImpl.java:888]: Use meaningful variable names.

File: components/appmgt/org.wso2.carbon.appmgt.impl/src/main/java/org/wso2/carbon/appmgt/impl/dao/AppMDAO.java (hunk 4 above)
Comment (lahiru823) [inline components/appmgt/org.wso2.carbon.appmgt.impl/src/main/java/org/wso2/carbon/appmgt/impl/dao/AppMDAO.java:4547]: This function is a private function and the SQL exception is thrown within the DAO layer

File: components/appmgt/org.wso2.carbon.appmgt.impl/src/main/java/org/wso2/carbon/appmgt/impl/APIProviderImpl.java (hunk 5 above)
Comment (lahiru823) [inline components/appmgt/org.wso2.carbon.appmgt.impl/src/main/java/org/wso2/carbon/appmgt/impl/APIProviderImpl.java:888]: Commit# https://github.com/wso2/carbon-appmgt/commit/b6127802564531ec04ce3a48f1c5b937aa676824

File: components/appmgt/org.wso2.carbon.appmgt.impl/src/main/java/org/wso2/carbon/appmgt/impl/APIProviderImpl.java (hunk 3 above)
Comment (lahiru823) [inline components/appmgt/org.wso2.carbon.appmgt.impl/src/main/java/org/wso2/carbon/appmgt/impl/APIProviderImpl.java:905]: Commit# https://github.com/wso2/carbon-appmgt/commit/a3e65c45d7048228694b38954493d368691304c0

File: components/appmgt/org.wso2.carbon.appmgt.impl/src/main/java/org/wso2/carbon/appmgt/impl/APIProviderImpl.java (hunk 2 above)
Comment (lahiru823) [inline components/appmgt/org.wso2.carbon.appmgt.impl/src/main/java/org/wso2/carbon/appmgt/impl/APIProviderImpl.java:892]: Commit# https://github.com/wso2/carbon-appmgt/commit/a346921d7ab0bffa073896e6b45a192eff977212

File: components/appmgt/org.wso2.carbon.appmgt.api/src/main/java/org/wso2/carbon/appmgt/api/model/JavaPolicy.java (hunk 1 above)
Comment (lahiru823) [inline components/appmgt/org.wso2.carbon.appmgt.api/src/main/java/org/wso2/carbon/appmgt/api/model/JavaPolicy.java:28]: Fixed with commit# https://github.com/lahiru823/carbon-appmgt/commit/ca07f3266be9d05c73d40023308ff203a24e3613

**Pull request merged**